base/cvd/cuttlefish/host/libs/zip/zip_cc.h

```diff
@@ -56,101 +62,154 @@ struct ZipStat {
   std::optional<ZipCompression> compression_method;
 };
 
-/* A read-only or read-write storage for data.
- *
- * Can be used to provide data to libzip with the `From` methods, and can also
- * be created to retrieve data out of libzip using `Zip::GetFile`.
- *
- * No methods to write data directly are implemented, but data can be written
- * using `Zip::CreateFromSource` and `AddFile`.
- */
-class ZipSource {
+class ReadableZipSource {
  public:
-  friend class Zip;
+  friend class ReadableZip;
+  friend class SeekableZipSource;
+  friend class WritableZip;
+  friend class WritableZipSource;
   friend class ZipSourceReader;
+  friend class SeekingZipSourceReader;
 
-  /* Read-only data access based on calling callback methods. */
-  static Result<ZipSource> FromCallbacks(
-      std::unique_ptr<SeekableZipSourceCallback>);
-  /* Read-write data access based on a copy of `data` up to `size`. Does not
-   * hold a reference to `data`. */
-  static Result<ZipSource> FromData(const void* data, size_t size);
-  /* Read-write data access to a file on disk. */
-  static Result<ZipSource> FromFile(const std::string& path);
-  /* Read-write data access to an in-memory buffer based on serializing a zip
-   * archive. */
-  static Result<ZipSource> FromZip(class Zip);
-
-  ZipSource(ZipSource&&);
-  ~ZipSource();
-  ZipSource& operator=(ZipSource&&);
+  static Result<ReadableZipSource> FromCallbacks(
+      std::unique_ptr<ReadableZipSourceCallback>);
+
+  // Can be safely called with a subclass type.
+  ReadableZipSource(ReadableZipSource&&);
+  virtual ~ReadableZipSource();
+  // Can be safely called with a subclass type.
+  ReadableZipSource& operator=(ReadableZipSource&&);
 
   Result<ZipStat> Stat();
 
   /* Returns a RAII instance that puts this instance in an "open for reading"
-   * state. Can fail. */
+   * state. Can fail. Should not outlive this instance. */
   Result<class ZipSourceReader> Reader();
 
  private:
   struct Impl;  // For pimpl: to avoid exposing libzip headers
 
-  ZipSource(std::unique_ptr<Impl>);
+  ReadableZipSource(std::unique_ptr<Impl>);
 
   std::unique_ptr<Impl> impl_;
 };
 
-/* A `ZipSource` in an "open for reading" state. */
+class SeekableZipSource : public ReadableZipSource {
+ public:
+  friend class ReadableZip;
+  friend class WritableZipSource;
+
+  static Result<SeekableZipSource> FromCallbacks(
+      std::unique_ptr<SeekableZipSourceCallback>);
+
+  SeekableZipSource(SeekableZipSource&&);
+  ~SeekableZipSource() override;
+  SeekableZipSource& operator=(SeekableZipSource&&);
+
+  /* Returns a RAII instance that puts this instance in an "open for reading"
+   * state. Can fail. Should not outlive this instance. */
+  Result<class SeekingZipSourceReader> Reader();
+
+ private:
+  SeekableZipSource(std::unique_ptr<Impl>);
+};
+
+class WritableZipSource : public SeekableZipSource {
```

Comment:
This inheritance implies that all `WritableZipSource`s are also readable, but that's not always true. For example, when writing zip-compressed data to a network stream.

Reply:
I don't believe libzip has the concept of a write-only `zip_source_t` instance.

https://libzip.org/documentation/zip_source_function_create.html

> Depending on the uses, there are three useful sets of commands to be supported by a `zip_source_callback`
>
> - read source
> - seekable read source
> - read/write source